man/man6/wpuzzle.6
.TH WPUZZLE 6 "Version 1, Release 4" "W Window System" "W PROGRAMS"
.SH NAME
wpuzzle \- sliding puzzle for W
.SH SYNOPSIS
.nf
wpuzzle [-n <size>] [-f <font family>] [-s <font size>]  [<bitmap file>]
.fi
.SH DESCRIPTION
The
.I wpuzzle
program displays a puzzle consisting of
.B N
times
.B N
squares, where
.B N
can be set via the
.B \-n
command line option.  One of the squares is blank.  The objective of the game
is to bring the puzzle into the solved state by successively sliding an
occupied square to the location of the blank one.  The puzzle is always
solvable.
.PP
The puzzle can either display an image (if the
.I bitmap file
argument is given; the file must be in GEM \.IMG format) or a sequence of
numbers (in the font defined via the
.B \-f
command line option).
.PP
This program was inspired by the
.I xpuzzle
client program for the X window system, which in turn was inspired by the
original "15-puzzle".  However,
.I wpuzzle
has been rewritten from scratch and contains no code for solving the puzzle
automatically.
.SH OPTIONS
.I wpuzzle
accepts the following options:
.TP
.B \-n \fIsize\fP
Sets the size of the puzzle, i.e., the number of squares along an edge.
.I Size
must be larger than 1.  The default
.I size
is 4.
.TP
.B \-f \fIfont\fP
Sets the font used when no
.I bitmap file
argument is given.  The default
.I font
is "cour10.wfnt".
.SH CONTROL
To slide one of the squares adjacent to the blank into the blank space, just
click on it.  You can also click on any of the squares in the same row or
column as the blank; this will move all the intervening squares in the same
direction.
.PP
To exit the program, press the 'q' key.
.SH AUTHORS
.I Wpuzzle
was written by Jens Kilian.  The code for reading \.IMG files was written
by Torsten Scherer, author of
.IR W (1).
The code for sliding multiple squares was written by Simon Kagedal.
.SH "SEE ALSO"
.IR wserver (1)
.SH BUGS
.I Wpuzzle
cannot solve the puzzle on its own.


wt/CHANGES
++oddie 9/2000:
- fixed listbox/filesel resizing & added minimum size setopt.

++bsittler 7/2000:
- Fixed vt color/reverse and attribute resetting
- Add HTTP/1.1 'Host' header to Wetscape requests
- Other minor changes (netgames, wvt, waudio)

++jcarlson 7/2000:
- WTINYSCREEN changes

++eero, 5/1999:
- Fixed key handling for special keys in Wt/getstring.c & wedit/wedit.c.

++eero, 1/1999:
- Added support for new W special keys so that now on W platforms which
implement special key mappings, you don't need to use Emacs shortcuts
with text widgets (or remap your console cursor keys)...


------ notes for version released with W R1PL0 ----------

ADDITIONAL WIDGETS

- Arrow, a simple button with an arrow drawn into it.

- Dial, a dial widget where you can position a needle into desired
  angle.  This isn't in the W toolkit library itself as then you would
  need to link math library to all of your Wt programs.  You'll find it
  in the warzone game tar-ball.

- Edittext, mainly intended for simple one page stuff, but as this
  spaghetti can do some additional stuff like paragraph formatting,
  there's a real abuse built on this, the WEdit.  As edittext widget is
  large and buggy, it's in the WEdit directory instead of the toolkit.

- Menu, a strip of buttons from which popups pop-up.  Both the popups
  and menus use a structure defined in Wt.h instead of strings that Kay
  was using.  The reason is that I want callbacks to follow
  automatically hierarchically opened popups.

- Popup, hierarchical popups.  Can include subpopups, images, lines and
  in addition items can also be disabled and checked.

- Range, a scrollbar composite with which you can select a value between
  given limits.

- Select, a getstring widget coupled with an arrow openable listbox from
  which items can be selected into a getstring widget.


OTHER CHANGES

- vt can now do everything wterm  can do + user can rebind the colors
  in wtrc.

- Added '.wtconfig' file and made makefiles use the W compilation
  configuration file '.config' in parent directory.

- wt_do_event() doesn't call wt_dispatch_event() anymore with
  EVENT_GADGET/GADGET_EXIT event as those might not have valid window
  pointer.  If you want to check for abnormal (unboud) exit, check the
  wt_return value.  GADGET_EXIT returns a negative number and
  GADGET_CLOSE (in shell widget) a positive number now.

- Added xterm mouse events, reverse mode, fd/bgcolor setting + improved
  double/triple click timing in VT widget.

- Aliases a couple new tags into others in HTML widget and added
  HTML header (ignored) and TITLE parsing.

- Widget event processing functions return now unrecognized or
  non-processed WEVENTs or NULL instead of long.  Assumption is that if
  function returns WEVENT, the widget whose action it was, is still
  around.  If in doubt, use NULL.  Unrecognized key events should be
  returned for focus change though.  (actually it could still have been
  an int, but returning a WEVENT offers a possibility for event mapping
  widgets.)

- Added wt_global structure containing Wt configuration variables and
  made widgets to use it.  wt_init() will now read these variables from
  ~/.wtrc + a list of (variable, value) pairs.  Widget specific
  initialization code in their init() functions may read these with the
  wt_variable() function.

- HTML widget fonts can be configured by above method.

- Added wt_loadfont() function which will use the user specified Wt
  default if widget doesn't specify some font parameter.  It also exits
  program if Wt default fonts can't be loaded, because otherwise
  unavailable fonts may chrash naive programs.

- Changed widgets to reflect the new w_loadfont() syntax (NULL/zero
  values tell Wlib/server to use server defaults).

- Added new WT_FONTSIZE define.  The earlier WT_FONT setopt sets now the
  font *family* name, new one sets the font size.

- Getopting WT_FONT returns WFONT* pointer also for above reason.

- Removed 'fontname' members from widget structures.  Needed information
  and more can be gotten from the 'font' member if needed.

- Fileselector has now separate lists for directory and file names.
  Both of course have an autolocator (type & return in addition to point
  & click).

- Wt_keys.h defines values for some special keys like cursors etc.  Half
  of the defines are Emacs style and half what the special keys itself
  give or are in use on other platforms.

- Changed Wt option defines.  There were a couple overlapping ones.


------- Kay's original notes: ----------------------------------------------

- added some missing close() and truncate the file to zero length before
  writing a new selection in selection.c.

- better selection handling in VT widget.

- made everything resizeable.

- applied patches by eero and fixed some more bugs myself.

- implemented a simple clipboard and changed getstring and vt widget to use it.

- fixed several serious bugs in the html widget.

- wrote VT widget.

- added an option to the icon editor to sync edit field with display
  field (simon).

- added missing special chars to html widget (simon).

- added code to the fileselector to set a default file name.

- lots of changes by eero (button.c, fileselect.c, listbox.c, getstring.c), some
  fixes to his changes by me.

- added missing wt_ungetfocus() to listbox and getstring.

- wrote Tk-like packer widget.

- new fileselect widget and testprogram by eero. Added better filename
  matching (eeros version was a joke, really).

- getstring and listbox updates by eero.

- changed the scrollbar widget's user interface to become more user friendly.
  See scrollbar.doc for details on how to use it. Also changed viewport and
  listbox to use the new interface.

- added listbox_focus(). Otherwise listbox cannot receive keyboard focus...

- big bugfix: when wt_delete() is called on a widget all pending timers must
  be canceled! (affected list, iconedit and getstring).


games/wpuzzle.c
/*
 * wpuzzle.c, a part of the W Window System
 *
 * Copyright (C) 1994-1997 by Jens Kilian, Torsten Scherer, Simon Kagedal
 *
 * This package is free software; you can redistribute it and/or modify it
 * under the terms specified in the docs/COPYRIGHTS file coming with this
 * package.
 *
 * -- an interactive puzzle game.  Inspired by an X client program.
 *
 *    Contributors:
 *	Jens Kilian	(original author)
 *	Torsten Scherer	(image reading functions)
 *	Simon Kagedal	(multiple moves)
 *
 *    $Log: wpuzzle.c,v $
 *    Revision 1.2  1999/05/16 13:09:55  eero
 *    add static keywords
 *
 *    Revision 1.1.1.1  1998/11/01 17:03:27  eero
 *    Entered W window system to CVS
 *
 *    Revision 1.5  1996/02/05  16:10:00  tesche
 *    w_readimg() moved into Wlib
 *
 *    Revision 1.4  1996/01/25  20:59:22  root
 *    Speed up multiple moves.
 *
 *    Revision 1.3  1996/01/24  17:47:12  root
 *    Fix that RCS ID ...
 *
 *    Revision 1.2  1996/01/24  17:38:16  root
 *    Added multiple-move enhancements by Simon Kagedal.
 *
 *    Revision 1.1  1996/01/24  17:12:52  root
 *    Initial revision
 */

static const char rcsid[] = "$Id: wpuzzle.c,v 1.2 1999/05/16 13:09:55 eero Exp $";

#include <Wlib.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>


/*  Makes a position in the puzzle field from coordinates. (/simon) */

#define	POS(x,y) (N*(y)+(x))


/*
 *  Allocate & initialize the puzzle.
 */

static int *init_puzzle(int N)
{
  int size = N*N;
  int i, *puzzle, x, y, m, n;
  
  /* Allocate memory. */
  
  if ((puzzle = malloc(sizeof(int)*size)) == 0) {
    return 0;
  }
  
  /* Initialize to solved state. */
  
  for (i = 0; i < size-1; ++i) {
    puzzle[i] = i+1;
  }
  puzzle[n = (size-1)] = 0;
  x = y = N-1;
  m = 0;
  
  /* Perform random moves. */
  
  srand(time(0));
  for (i = 0; i < N*N*100;) {
    switch (rand() & 3) {
      
    case 0:
      if (y > 0) m = N*--y + x; else continue;
      break;
      
    case 1:
      if (x > 0) m = N*y + --x; else continue;
      break;
      
    case 2:
      if (x<N-1) m = N*y + ++x; else continue;
      break;
      
    case 3:
      if (y<N-1) m = N*++y + x; else continue;
      break;
    }
    
    puzzle[n] = puzzle[m];
    puzzle[n = m] = 0;
    ++i;
  }
  
  return puzzle;
}

/*
 *  Draw the bitmap version of the puzzle.
 */

static void bitmap_puzzle(WWIN *window, WWIN *bm, int bm_width, int bm_height,
			int N, int x_size, int y_size, int *puzzle)
{
  int i, size = N*N;
  int xoff = (N*x_size - bm_width)/2;
  int yoff = (N*y_size - bm_height)/2;

  /* Clear the window. */

  w_setmode(window, M_CLEAR);
  w_pbox(window, 0, 0, N*x_size, N*y_size);
  w_setmode(window, M_DRAW);

  /* Draw the cells. */

  for (i = 0; i < size; ++i) {
    int x0 = (i % N)*x_size;
    int y0 = (i / N)*y_size;
    
    if (puzzle[i]) {
      int x1 = ((puzzle[i]-1) % N)*x_size - xoff;
      int y1 = ((puzzle[i]-1) / N)*y_size - yoff;
      int width = x_size, height = y_size;

      if (x1 < 0) {
	width += x1;
	x0 -= x1;
	x1 = 0;
      } else if (x1+width > bm_width) {
	width = bm_width - x1;
      }
      if (y1 < 0) {
	height += y1;
	y0 -= y1;
	y1 = 0;
      } else if (y1+height > bm_height) {
	height = bm_height - y1;
      }
      w_bitblk2(bm, x1, y1, width, height, window, x0, y0);

    } else {
      w_pbox(window, x0, y0, x_size, y_size);
    }
  }
}

/*
 *  Draw the text version of the puzzle.
 */

static void text_puzzle(WWIN *window, WFONT *wfont, int N,
			int x_size, int y_size, int *puzzle)
{
  int i, size = N*N, width;
  char line[64];
  
  /* Clear the window. */
  
  w_setmode(window, M_CLEAR);
  w_pbox(window, 0, 0, N*x_size, N*y_size);
  w_setmode(window, M_DRAW);

  /* Draw the cells. */

  w_setfont(window, wfont);
  
  for (i = 0; i < size; ++i) {
    int x0 = (i % N)*x_size;
    int y0 = (i / N)*y_size;
    
    if (puzzle[i]) {
      w_box(window, x0, y0, x_size, y_size);
      sprintf(line, "%d", puzzle[i]);
      width = w_strlen(wfont, line);
      w_printstring(window,
		    x0 + (x_size-width)/2,
		    y0 + (y_size-wfont->height)/2,
		    line);
    } else {
      w_pbox(window, x0, y0, x_size, y_size);
    }
  }
}

/*
 *  Main event loop.
 */

static void do_puzzle(WWIN *window, int N, int x_size, int y_size, int *puzzle)
{
  WEVENT *ev;
  int stop = 0;
  int x, y, dx, dy, l, i, from_x, from_y, to_x, to_y, width, height;
  
  while (!stop) {
    ev = w_queryevent(NULL, NULL, NULL, -1);

    switch (ev->type) {

    case EVENT_MPRESS:
      x = ev->x / x_size;
      y = ev->y / y_size;

      /* Determine if move is legal and which direction it takes. */

      dx = dy = l = 0;

      for (i = 0; i < N; ++i) {
	if ((!puzzle[POS(x,i)]) && (i != y)) { /* Vertical move? */
	  l  = (y > i) ? y - i : i - y;
	  dy = (y > i) ? -1 : 1;
	  break;
	}
	if ((!puzzle[POS(i,y)]) && (i != x)) { /* Horizontal move? */
	  l  = (x > i) ? x - i: i - x;
	  dx = (x > i) ? -1 : 1;
	  break;
	}
      }

      if (l) {
	/* Perform the move. */
	
	for (i = l; i > 0; --i) {
	  from_x = (to_x = x + (dx*i)) - dx;
	  from_y = (to_y = y + (dy*i)) - dy;

	  puzzle[POS(to_x,to_y)] = puzzle[POS(from_x,from_y)];
	  puzzle[POS(from_x,from_y)] = 0;
	}

	/* Animate. */

	x = x_size * (x - (dx<0)*(l-1));
	y = y_size * (y - (dy<0)*(l-1));
	width  = x_size * (dx ? l : 1);
	height = y_size * (dy ? l : 1);

	for (i = (dx ? x_size : y_size); i > 0; --i) {
	  w_bitblk(window, x, y, width, height, x+dx, y+dy);
	  if (dx) {
	    w_vline(window, (dx<0) ? x+width-1 : x, y, y+height-1);
	  } else {
	    w_hline(window, x, (dy<0) ? y+height-1 : y, x+width-1);
	  }
	  x += dx;
	  y += dy;
	}

	/* Check if the puzzle is solved. */
	
	for (i = 0; i < N*N-1; ++i) {
	  if (puzzle[i] != i+1) {
	    break;
	  }
	}
	if (i == N*N-1) {
	  w_settitle(window, "SOLVED!");
	}
      } else {
	w_beep();
      }
      break;

    case EVENT_KEY:
      if ((ev->key & 0xff) == 'q') {
	stop = 1;
      }
      break;

    case EVENT_GADGET:
      if ((ev->key == GADGET_EXIT) ||
	  (ev->key == GADGET_CLOSE)) {
	stop = 1;
      }
      break;
    }
  }
}

/*
 *  Main program - window system initialization & termination.
 */

static void usage(char *progname)
{
  fprintf(stderr, "Usage: %s [-n size] [-f font family] [-s font size] [IMG file]\n", progname);
}

int main(int argc, char **argv)
{
  char *fname = NULL;	/* use defaults */
  short fsize = 0;
  int N = 4;
  int *puzzle;
  char line[64];
  int x_size, y_size;
  int i;
  char opt, *arg;
  BITMAP *bm;
  short bm_width, bm_height;
  WSERVER *wserver;
  WFONT *wfont;
  WWIN *window, *image = NULL;
  
  /* Parse command line arguments. */
  
  i = 1;
  while ((i < argc) && (argv[i][0] == '-')) {
    opt = argv[i][1];
    if (argv[i][2] != '\0') {
      arg = &argv[i][2];
    } else if (++i < argc) {
      arg = argv[i];
    } else {
      usage(argv[0]);
      return 1;
    }

    switch (opt) {

    case 'n':                /* set size of puzzle */
      N = strtol(arg, &arg, 10);
      if ((N < 2) || (*arg != '\0')) {
	usage(argv[0]);
	return 1;
      }
      break;

    case 'f':                /* set font family */
      fname = arg;
      break;

    case 's':                /* set font size */
      fsize = atoi(arg);
      break;

    default:
      usage(argv[0]);
      return 1;
    }

    ++i;
  }

  if (i < argc) {
    /* Try to load bitmap file. */
    if ((bm = w_readimg(argv[i], &bm_width, &bm_height)) == NULL) {
      return 1;
    } else if ((bm_width < N) || (bm_height < N)) {
      fprintf(stderr, "%s: image too small, must be at least %dx%d\n",
	      argv[0], N, N);
      w_freebm(bm);
      return 1;
    }
  } else {
    bm = 0;
  }

  /* Set up connection to server. */
  
  if ((wserver = w_init()) == 0) {
    fprintf(stderr, "%s: can't connect to wserver\n", argv[0]);
    return 1;
  }
  
  /* Calculate cell size. */
  
  if (bm) {
    wfont = 0;
    x_size = (bm_width+N-1)/N;
    y_size = (bm_height+N-1)/N;

  } else {
    if ((wfont = w_loadfont(fname, fsize, 0)) == 0) {
      fprintf(stderr, "%s: can't load font '%s%d'\n", argv[0], fname, fsize);
      return 1;
    }
    
    sprintf(line, "%d", N*N-1);
    x_size = y_size = w_strlen(wfont, line) + 8;
  }

  /* Try to create the window(s), then initialize & open it. */
  
  window = w_create(N*x_size, N*y_size, (W_MOVE|W_TITLE|W_CLOSE|EV_MOUSE|EV_KEYS));
  if (window == 0) {
    fprintf(stderr, "%s: can't create window\n", argv[0]);
    return 1;
  }
  w_settitle(window, "WPuzzle");

  if(bm) {
    if (!(image = w_create(bm_width, bm_height, 0))) {
      fprintf(stderr, "%s: can't create bitmap\n", argv[0]);
      return 1;
    }
    w_putblock(bm, image, 0, 0);
    w_freebm(bm);
  }
  
  if(w_open(window, UNDEF, UNDEF) < 0) {
    fprintf(stderr, "%s: can't open window\n", argv[0]);
    return 1;
  }
  
  /* Initialize the puzzle. */
  
  if ((puzzle = init_puzzle(N)) == 0) {
    fprintf(stderr, "%s: can't allocate memory\n", argv[0]);
    return 1;
  }

  /* Draw the puzzle. */

  if (image) {
    bitmap_puzzle(window, image, bm_width, bm_height, N, x_size, y_size, puzzle);
    w_delete(image);
  } else {
    text_puzzle(window, wfont, N, x_size, y_size, puzzle);
  }
		  
  /* Play the game. */
  
  do_puzzle(window, N, x_size, y_size, puzzle);
  
  /* Exit. */
  
  w_close(window);
  w_delete(window);
  w_exit();
  return 0;
}
